homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,253 @@
+
+    connection.send_result(msg["id"])
+
+
+class UpdatePlatformProtocol(Protocol):
+    """Define the format that update platforms can have."""
+
+    async def async_list_updates(self, hass: HomeAssistant) -> list[UpdateDescription]:
+        """List all updates available in the integration."""
+
+    async def async_perform_update(
+        self,
+        hass: HomeAssistant,
+        identifier: str,
+        version: str,
+        **kwargs: Any,
+    ) -> bool:
+        """Perform an update."""
+
+
+@dataclasses.dataclass()
+class UpdateDescription:
+    """Describe an update update."""
+
+    identifier: str
+    name: str
+    current_version: str
+    available_version: str
+    changelog_content: str | None = None
+    changelog_url: str | None = None
+    icon_url: str | None = None
+    provides_backup: bool = False
+
+
+class UpdateManager:
+    """Update manager for the update integration."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the update manager."""
+        self._hass = hass
+        self._store = storage.Store(
+            hass=hass,
+            version=STORAGE_VERSION,
+            key=DOMAIN,
+        )
+        self._skip: set[str] = set()
+        self._platforms: dict[str, UpdatePlatformProtocol] = {}
+        self._loaded = False
+
+    def add_platform(
+        self,
+        integration_domain: str,
+        platform: UpdatePlatformProtocol,
+    ) -> None:
+        """Add a platform to the update manager."""
+        self._platforms[integration_domain] = platform
+
+    async def gather_updates(self) -> list[dict[str, Any]]:
+        """Gather updates."""
+        if not self._loaded:
+            self._skip = set(await self._store.async_load() or [])
+            self._loaded = True
+
+        updates: dict[str, list[UpdateDescription] | None] = {}
+
+        for domain, update_descriptions in zip(
+            self._platforms,
+            await asyncio.gather(
+                *(
+                    self._get_integration_info(integration_domain, registration)
+                    for integration_domain, registration in self._platforms.items()
+                )
+            ),
+        ):
+            updates[domain] = update_descriptions
+
+        return [
+            {
+                "domain": integration_domain,
+                **dataclasses.asdict(description),
+            }
+            for integration_domain, update_descriptions in updates.items()
+            if update_descriptions is not None
+            for description in update_descriptions
+            if f"{integration_domain}_{description.identifier}_{description.available_version}"
+            not in self._skip
+        ]
+
+    def domain_is_valid(self, domain: str) -> bool:
+        """Return if the domain is valid."""
+        return domain in self._platforms
+
+    @callback
+    def _data_to_save(self) -> list[str]:
+        """Schedule storing the data."""
+        return list(self._skip)
```

Comment:
For future expansion, let's make this a dictionary. That way if we want to add a field, we don't have to change the root data format and remain backwards compatible.

```suggestion
        return {"skipped": list(self._skip)}
```

For example, we might want to store an update history.